drools-parser/src/main/antlr4/org/drools/parser/DRLLexer.g4

```diff
@@ -114,3 +114,8 @@ fragment
 DrlUnicodeEscape
     :   '\\' 'u' HexDigit HexDigit HexDigit HexDigit
     ;
+
+mode RHS;
+RHS_WS : [ \t\r\n\u000C]+ -> channel(HIDDEN);
+DRL_END : 'end' -> popMode;
+RHS_CHUNK : ~[ \t\r\n\u000C]+ ;
```

Comment:
I have introduced Island mode because it's easier to implement this PR. `RHS_CHUNK` is separated by white-space etc. so RHS would include multiple `RHS_CHUNK`s. I was not able to define a Token to represent single whole RHS text because such a Token greedy consumes `end`. If you have an idea to define a better Lexer/Token, please let me know, @tarilabs .  But this `RHS_CHUNK` doesn't seem to have an issue for now.

Reply:
I concur with your analysis and approach in general, good job!!

The only thing which I don't know, so sharing my doubt here, is that I'm afraid that there is some WS-sensitivity in the case of RHS for termination.
I'm specifically referring to the `DRL_END : 'end'` I'm not sure is enough for all situations, for instance in the case of:

```
rule R1
...
then
  System.out.println($s.endsWith("xyz"));
end
```

I'm not sure if this would pop at the "`end`" in the "`endsWith()`".

IFF that is actually a problem, you could consider two strategies:

Strategy A:
could solve on lexing side, you could have something ~like:

```
mode RHS;
RHS_CHUNK : anything except drl_end;
DRL_END : WS_OR_NL+ 'end' WS_OR_NL+ -> popMode;
```

not working, of course :) just to explain the idea: you are expecting anything (including newline, whitespace, etc) until you reach a line that contains only `'end'`.

WS_OR_NL could be something like `\p{Whitespace}` if that Unicode class contains newlines too --not sure, worthy to check.

Strategy B:
could solve by a mix of lexer and parser rule (if above fails).

I think worthy to have a test with the example rule above, and see if the alleged issue which could potentially arise, does it actually happens or not

Separately, in general, I believe it is fine to have "multiple" RHS_CHUNKs etc, that is not a problem.

Reply:
Thank you, @tarilabs ! Sure, I'll add a test and address it.

Reply:
`System.out.println($s.endsWith("xyz"));` has no problem because RHS mode already parses with whitespace delimiters.

However, RHS like
```
then
   int end = 10;
end
```
has a problem.

Indeed, the current DRL parser checks  a token next to `end`, then only when the next tokens are one of `rule`, `query`, `function` etc. (= the top level elements), it considers the end of RHS. 

https://github.com/kiegroup/drools/blob/main/drools-drl-parser/src/main/java/org/drools/drl/parser/lang/DRL6Parser.java#L4626
https://github.com/kiegroup/drools/blob/main/drools-drl-parser/src/main/java/org/drools/drl/parser/lang/ParserHelper.java#L248-L255

I followed the same approach. Could you share your thoughts on this? Thanks!

Reply:
I think 0734bab is a valid approach, and btw thanks for having checked the use-cases of "end" in a RHS block!

On other hand, I wonder if it's just enough to have something ~like:

```
DRL_END : 'end' NL+ EOF+;
```

It would avoid the logic of having to chain-together `end` with the next "top level" DRL-keyword.
That would simplify without the work needed in 0734bab ?
I hope I was able to explain myself 😅

Reply:
@tarilabs Thank you! Yes, I'll try the one.

Reply:
Thank you @tarilabs , I implemented it and it's much simpler. Now we assume that `end` has newline or EOF.  It's not clearly stated in docs, but we may assume that people don't write multiple rules in one line.

https://docs.drools.org/latest/drools-docs/drools/language-reference/index.html#con-drl_drl-rules

Reply:
> Now we assume that end has newline or EOF. It's not clearly stated in docs, but we may assume that people don't write multiple rules in one line.

I think it's a good assumption for now 👍 
And we can always make it more explicit if we finalize for this solution.

Another thing you might want to consider, iirc (@mariofusco may confirm here) is that ConsequenceDescr keeps the original indenting/whitespace.
You can consider using something like:

https://github.com/kiegroup/drools-ansible-rulebook-integration/blob/e21a53d42d7a6d37fd72f11c028d1a24c7480076/drools-ansible-rulebook-integration-protoextractor/src/main/java/org/drools/ansible/rulebook/integration/protoextractor/ast/AbstractNode.java#L43-L47

for the `RHS_CHUNK`s, it should allow you to extract the original text from start to end, including whitespace, tab, etc.
(if you are not doing that already)

Reply:
> ConsequenceDescr keeps the original indenting/whitespace.

Thanks, @tarilabs and yes, I have done the same:

https://github.com/kiegroup/drools-lsp/blob/main/drools-parser/src/main/java/org/drools/parser/DRLVisitorImpl.java#L156
https://github.com/kiegroup/drools-lsp/blob/main/drools-parser/src/main/java/org/drools/parser/ParserStringUtils.java#L25-L35

`RHS_CHUNK` is just used for convenience in lexer/parser rules.